src/Libraries/Microsoft.Extensions.AI/ToolGrouping/ToolGroupingChatClient.cs

```diff
@@ -0,0 +1,423 @@
+
+    /// <summary>
+    /// Processes a set of function call expansion requests and determines which chat messages to add and whether
+    /// expansion should terminate.
+    /// </summary>
+    /// <remarks>
+    /// Expansion requests that exceed the maximum allowed iterations or specify invalid or duplicate
+    /// groups are ignored. Only one new group can be expanded per call; subsequent requests for the same group are
+    /// treated as duplicates. The expanded group reference is updated only if a new group is successfully
+    /// expanded.
+    /// </remarks>
+    /// <param name="expansionRequests">A list of expansion requests to process. Each request specifies a function call and its associated arguments.</param>
+    /// <param name="expandedGroup">A reference to the tool group that has been expanded. This parameter is updated if a new group is expanded during processing.</param>
+    /// <param name="expansionIterationCount">The current number of expansion iterations performed for this request. Used to enforce the maximum allowed expansions per request.</param>
+    /// <returns>
+    /// A tuple containing a list of chat messages to add in response to the expansion requests, and a boolean value
+    /// indicating whether expansion should terminate. If no valid expansions are processed, the message list will be
+    /// empty and termination will be signaled.
+    /// </returns>
+    private (IList<ChatMessage> messagesToAdd, bool shouldTerminate) ProcessExpansions(
+        List<FunctionCallContent> expansionRequests,
+        ref AIToolGroup? expandedGroup,
+        int expansionIterationCount)
+    {
+        if (expansionRequests.Count == 0)
+        {
+            return (messagesToAdd: [], shouldTerminate: true);
+        }
+
+        var didExpandNewGroup = false;
+        var maxExpansionsReached = expansionIterationCount >= _options.MaxExpansionsPerRequest;
+        var contents = new List<AIContent>(expansionRequests.Count);
+
+        foreach (var expansionRequest in expansionRequests)
+        {
+            if (maxExpansionsReached)
+            {
+                contents.Add(new FunctionResultContent(
+                    callId: expansionRequest.CallId,
+                    result: "Max expansion iteration count reached; ignoring additional expansion request."));
+                continue;
+            }
+
+            if (expansionRequest.Arguments is not { Count: > 0 } arguments ||
+                !arguments.TryGetValue(GroupNameParameter, out var groupNameArg) ||
+                groupNameArg is null)
+            {
+                contents.Add(new FunctionResultContent(
+                    callId: expansionRequest.CallId,
+                    result: "No group name was specified; ignoring expansion request."));
+                continue;
+            }
+
+            var groupName = groupNameArg.ToString();
+            if (groupName is null || !_toolGroupsByName.TryGetValue(groupName, out var group))
```

Comment:
So we always return all Tools in a group, even if those tools were not part of the options for this request?   I can see how that might make sense - we have to choose a behavior - are the groupings additive to call, or does the call filter the groupings.  Given groupings are new and the call behavior exists, I had assumed the call would filter the groupings (but I wouldn't trust my perspective here 😄).

Reply:
Yeah, we could definitely treat the group as a "filter" more than an extra set of available tools if we wanted. Maybe this is another reason to define groups directly on `ChatOptions` rather than configuring them on the middleware.